src/test/java/uk/gov/pay/directdebit/webhook/gocardless/services/WebhookGoCardlessServiceTest.java

```diff
@@ -0,0 +1,109 @@
+        verify(mockedGoCardlessService).storeEvent(secondGoCardlessEvent);
+    }
+
+    @Test
+    public void shouldStoreEventsWithAnUnhandledResource() {
+        GoCardlessEvent goCardlessEvent = aGoCardlessEventFixture().withResourceType(UNKNOWN).withAction("created").toEntity();
+
+        List<GoCardlessEvent> events = Collections.singletonList(goCardlessEvent);
+        webhookGoCardlessService.handleEvents(events);
+        verify(mockedGoCardlessService).storeEvent(goCardlessEvent);
+        verifyZeroInteractions(mockedTransactionService);
+    }
+
+    @Test
+    public void shouldStoreButNotHandlePaymentEventsWithAnInvalidAction() {
+        GoCardlessEvent goCardlessEvent = aGoCardlessEventFixture().withResourceType(PAYMENTS).withAction("not_handled").toEntity();
+
+        List<GoCardlessEvent> events = Collections.singletonList(goCardlessEvent);
+        webhookGoCardlessService.handleEvents(events);
+        verify(mockedGoCardlessService).storeEvent(goCardlessEvent);
+        verifyZeroInteractions(mockedTransactionService);
+    }
+
+    @Test
+    public void shouldStoreAndHandlePaymentEventsWithAValidAction() {
+        GoCardlessEvent goCardlessEvent = aGoCardlessEventFixture().withResourceType(PAYMENTS).withAction("paid_out").toEntity();
+
+        PaymentRequestEvent paymentRequestEvent = PaymentRequestEvent.paidOut(transaction.getPaymentRequestId());
+        when(mockedGoCardlessService.findPaymentForEvent(goCardlessEvent)).thenReturn(goCardlessPayment);
+        when(mockedTransactionService.findTransactionFor(goCardlessPayment.getTransactionId())).thenReturn(transaction);
+        when(mockedTransactionService.paidOutFor(transaction)).thenReturn(paymentRequestEvent);
+
+        List<GoCardlessEvent> events = Collections.singletonList(goCardlessEvent);
+        webhookGoCardlessService.handleEvents(events);
+        verify(mockedGoCardlessService).storeEvent(goCardlessEvent);
+        verify(mockedTransactionService).paidOutFor(transaction);
+    }
+
+    @Test
+    public void shouldStoreAndHandleMandateEventsWithAValidAction() {
+        GoCardlessEvent goCardlessEvent = aGoCardlessEventFixture().withResourceType(MANDATES).withAction("created").toEntity();
+
+        PaymentRequestEvent paymentRequestEvent = PaymentRequestEvent.mandateCreated(transaction.getPaymentRequestId());
+        when(mockedGoCardlessService.findMandateForEvent(goCardlessEvent)).thenReturn(goCardlessMandate);
+        when(mockedTransactionService.findTransactionForMandateId(goCardlessMandate.getMandateId())).thenReturn(transaction);
+        when(mockedTransactionService.mandateCreatedFor(transaction)).thenReturn(paymentRequestEvent);
+
+        List<GoCardlessEvent> events = Collections.singletonList(goCardlessEvent);
+        webhookGoCardlessService.handleEvents(events);
+        verify(mockedGoCardlessService).storeEvent(goCardlessEvent);
+        verify(mockedTransactionService).mandateCreatedFor(transaction);
+    }
+}
```

Comment:
Do we also want a `shouldStoreButNotHandleMandateEventsWithAnInvalidAction`?

Reply:
👍